src/garage/torch/modules/gaussian_mlp_module.py

```diff
@@ -0,0 +1,255 @@
+        uncentered_log_std = torch.zeros(*broadcast_shape) + self._init_std_param
+
+        return mean, uncentered_log_std
+
+
+class GaussianMLPAdaptiveModule(GaussianMLPBaseModule):
+    """
+    GaussianMLPModule which has two different mean and std network.
+    """
+    def __init__(self,
+                 input_dim,
+                 output_dim,
+                 **kwargs):
+        super(GaussianMLPAdaptiveModule, self).__init__(input_dim, output_dim, **kwargs)
+
+        self._mean_module = MLPModule(
+            input_dim=self._input_dim,
+            output_dim=self._action_dim,
+            hidden_sizes=self._hidden_sizes,
+            hidden_nonlinearity=self._hidden_nonlinearity,
+            hidden_w_init=self._hidden_w_init,
+            hidden_b_init=self._hidden_b_init,
+            output_nonlinearity=self._output_nonlinearity,
+            output_w_init=self._output_w_init,
+            output_b_init=self._output_b_init,
+            layer_normalization=self._layer_normalization)
+
+        self._log_std_module = MLPModule(
+            input_dim=self._input_dim,
+            output_dim=self._action_dim,
+            hidden_sizes=self._std_hidden_sizes,
+            hidden_nonlinearity=self._std_hidden_nonlinearity,
+            hidden_w_init=self._std_hidden_w_init,
+            hidden_b_init=self._std_hidden_b_init,
+            output_nonlinearity=self._std_output_nonlinearity,
+            output_w_init=self._std_output_w_init,
+            output_b_init=self._init_adaptive_b,
+            layer_normalization=self._layer_normalization)
+
+    def _init_adaptive_b(self, b):
+        return nn.init.constant_(b, self._init_std_param.item())
+
+    def _get_mean_and_log_std(self, inputs):
+        return self._mean_module(inputs), self._log_std_module(inputs)
+
+
+class GaussianMLPSharedModule(GaussianMLPBaseModule):
+    """
+    GaussianMLPModule which has only one mean network.
+    Std will be a parameter
+    """
+    def __init__(self,
+                 input_dim,
+                 output_dim,
+                 **kwargs):
+        super(GaussianMLPSharedModule, self).__init__(input_dim, output_dim, **kwargs)
+
+        self._shared_mean_log_std_network = MLPModule(
+            input_dim=self._input_dim,
+            output_dim=self._action_dim * 2,
+            hidden_sizes=self._hidden_sizes,
+            hidden_nonlinearity=self._hidden_nonlinearity,
+            hidden_w_init=self._hidden_w_init,
+            hidden_b_init=self._hidden_b_init,
+            output_nonlinearity=self._output_nonlinearity,
+            output_w_init=self._output_w_init,
+            output_b_init=self._init_shared_b,
+            layer_normalization=self._layer_normalization
+        )
+
+    def _init_shared_b(self, b):
+        nn.init.zeros_(b[..., :self._action_dim])
+        nn.init.constant_(b[..., self._action_dim:], self._init_std_param.item())
+
+    def _get_mean_and_log_std(self, inputs):
```

Comment:
can you actually implement this as a two-headed MLP? the TF implementation (which i wrote) isn't very nice.

basically what you want is an MLP module with `len(hidden_sizes)` hidden layers, but 2 different output layers. you can implement `MultiHeadedMLPModule` to do this. it will take in a parameter `n_heads` and return `n` outputs, with an output layer for each n.

this will be really useful in the future for many architectures.

--then you can just do
```python
def __init__(self, ...):
    self._shared_mean_log_std_network =  MultiHeadedMLPModule(n_heads=2, ...)

def _get_mean_and_log_std(self, input):
    return self._shared_mean_log_std_network(input)
```

Reply:
the topology for `n` hidden layers and `k` output heads is
```
[input] --> [hidden_1] --> ... --> [hidden_n] --> | --> [output_head_1]
                                                  |...
                                                  | --> [output_head_k]
```

note that you need to support a different `output_b_init` and `output_w_init` for each output head. that's pretty easy -- just have inputs `output_b_inits` and `output_w_inits`. for simplicity, you can still support `output_b_init`/`output_w_init` and use that for all heads (but only allow the user to specify one at at time)

Reply:
Implemented MultiHeadedMLPModule